### ISSUE TITLE
Function to round a value to a precision rather than an integer?

### ISSUE THREAD

**Issue by blixt** [sass/sass#1847]
I propose a rounding function that is better at handling decimal numbers. Basically, something that mirrors JavaScript's `Number.prototype.toPrecision` function:

``` javascript
var a = 0.16, b = 1.6;
a.toPrecision(1) // gives 0.2
b.toPrecision(1) // gives 2
```

Just using `round` in the two above cases would give `0` and `2` instead.

**Comment by cimmanon** [comment]
You can already do this yourself.  Don't know that this needs to be part of the standard library.

``` scss
@function to-precision($val, $places) {
    $x: pow(10, $places);
    @return round($val * $x) / $x;
}

.foo {
  bar: to-precision(10.55555, 2);
}
```

Note that this requires a function for determining the power of something.  Compass provides one and so do a few other libraries.

**Comment by blixt** [comment]
I think it belongs in a standard library that allows decimal operations such as division or `random()` (which returns 0..1 with up to five decimal points) as a way to avoid very long floating point numbers in the output.

Here's the full code for your suggested `to-precision` function, in case anyone wants to go with the homebrew version:

``` sass
@function pow($number, $exp) {
  $value: 1;
  @if $exp > 0 {
    @for $i from 1 through $exp {
      $value: $value * $number;
    }
  }
  @else if $exp < 0 {
    @for $i from 1 through -$exp {
      $value: $value / $number;
    }
  }
  @return $value;
}

@function to-precision($val, $places) {
    $x: pow(10, $places);
    @return round($val * $x) / $x;
}

.foo {
  bar: to-precision(10.55555, 2);
}
```

**Comment by chriseppstein** [comment]
Sass provides a global precision setting. What is your use case for a value-specific precision that differs from the global one?

**Comment by blixt** [comment]
@chriseppstein Interesting, I just tried `--precision` and that has the semantics I was looking for. However, is it possible to limit the setting within a block or some such?

My use case (which admittedly isn't critical to get the job done) is that I would like to do `precision(random() * $something, 1)` instead of `round(random() * $something * 10) / 10` because it's more readable and it's also cleaner in the sense that it will generate `0.5px` and `1px` instead of `0.5px` and `1.4px`. The fact that I care less about the fraction for numbers >= 1 is perhaps not very common.

**Comment by chriseppstein** [comment]
> The fact that I care less about the fraction for numbers >= 1 is perhaps not very common.

@blixt I agree. I think a custom function that does the rounding to your desired `>1` precision is best for this.

**Comment by stevenvachon** [comment]
I think that a second argument for `round()` makes sense for this. Less.js offers such. Custom functions add complexity around a standard documentation (like Sass'). Compass is useless to me and is just bloat.

**Comment by chriseppstein** [comment]
This is complicated because `round()` returns a number and Sass will still round that number to the global precision setting unless we somehow set that number with a flag that says "this instance of this number has this particular precision." Alternatively it can return a string but that number could no longer be used for further math calculations. @nex3 is against the idea of flagging a number to have an instance specific precision and strings create other complications that will end up creating headaches for people. So basically, we don't have a consensus a good solution to this issue is.

**Comment by stevenvachon** [comment]
A second argument would override the global default. Does Less.js solve any of the string stuff in its `round()`?

**Comment by nex3** [comment]
One possibility would be to add a `$precision` parameter to `inspect()` that would control the precision of any numbers emitted while converting the value to a string. Tying it to `inspect()` makes it very clear that you're getting a string, not a number, out the other end.

**Comment by whaaaley** [comment]
I like @stevenvachon's idea. Seems simple from a user's perspective. Plus it has more use than simply limiting decimals for output. I can't think of any scenario where I had wanted to round to the nearest whatever decimal and wanted it to be a string. If that were the case I'd imagine there would be lots of people casting strings to numbers for more math.

**Comment by chriseppstein** [comment]
@nex3 The semantics of inspect is that it is for debugging. I don't want to confuse that by telling people inspect is sometimes how you put a css value into css output. I'd rather just make an entirely new function for this.

But tbh, my pref is still to just set an instance level precision for the numbers where it matters and let the sass numeric output code do it's thing.

**Comment by stevenvachon** [comment]
@charlesroper via `precision()`? I think it looks too cryptic and is not self-documenting.

**Comment by whaaaley** [comment]
If self-documenting is the goal, how about `place-value()` or `round-to-place-value()` or maybe just `round-to()`...? I still like the idea of it being added to `round()`, because that's exactly that's what it's doing, it's just more specific.

**Comment by stevenvachon** [comment]
@dustindowell22 I think that's too wordy. `round()` is pretty clear, I think. But if necessary, what do you think of `decimal-round()` ?

**Comment by whaaaley** [comment]
@stevenvachon I totally agree. Its too long and round is clear.. `decimal-round()` is fine, but since this is sort of a variant of rounding I would go with `round-decimal()` instead.

**Comment by nex3** [comment]
I'm still against setting instance-specific flags if we can avoid it. If we don't want to use `inspect()`, I'd rather add a function called `to-css()` or `to-string()` that explicitly converts its argument into a string containing a plain-CSS value.

**Comment by stevenvachon** [comment]
Our opinions are on opposite poles and I can't think of a middle ground.
